plonk/src/proof_system/mod.rs

```diff
@@ -19,8 +21,10 @@ pub(crate) mod verifier;
 use crate::transcript::PlonkTranscript;
 pub use snark::PlonkKzgSnark;
 
-/// An interface for SNARKs.
-pub trait Snark<E: PairingEngine> {
+// TODO: (alex) should we name it `PlonkishSNARK` instead? since we use
```

Comment:
Wait, do you suggest we use associated types instead of trait bound, this could be difficult because our code are generic over `trait PlonkTranscript` right?

I saw your usage of passing in a `transcript: &mut IOPTranscript<E::Fr>,` in HP code, but I'm not sure if it's the best approach.
I think [marlin's approach](https://github.com/arkworks-rs/marlin/blob/master/src/lib.rs#L70) of being generic over `trait FiatShamirRng` is not bad

Reply:
> Wait, do you suggest we use associated types instead of trait bound, this could be difficult because our code are generic over trait PlonkTranscript right?

Well, i think it may actually simplify things. At the moment we have
``` rust
    /// Compute a SNARK proof of a circuit `circuit`, using the corresponding
    /// proving key `prove_key`. The witness used to
    /// generate the proof can be obtained from `circuit`.
    fn prove<C, R, T>(
        rng: &mut R,
        circuit: &C,
        prove_key: &Self::ProvingKey,
        extra_transcript_init_msg: Option<Vec<u8>>,
    ) -> Result<Self::Proof, Self::Error>
    where
        C: Arithmetization<E::Fr>,
        R: CryptoRng + RngCore,
        T: PlonkTranscript<E::Fq>;
```
and after the change we can use

``` rust
    /// Compute a SNARK proof of a circuit `circuit`, using the corresponding
    /// proving key `prove_key`. The witness used to
    /// generate the proof can be obtained from `circuit`.
    fn prove<C, R>(
        rng: &mut R,
        circuit: &C,
        prove_key: &Self::ProvingKey,
        extra_transcript_init_msg: Option<Vec<u8>>,
    ) -> Result<Self::Proof, Self::Error>
    where
        C: Arithmetization<E::Fr>,
        R: CryptoRng + RngCore,
```
The catch is we need to define 
```
Self::Transcript: PlonkTranscript
```
(for non-plonk SNARKs, the implementer may need to implement `PlonkTranscript` for a null struct so yes it simplifies things for us but complicates for others LOL.)


> I think [marlin's approach](https://github.com/arkworks-rs/marlin/blob/master/src/lib.rs#L70) of being generic over trait FiatShamirRng is not bad

I think whether generic over `Rng` or `Transcript` is orthogonal to the discussion. Either way, it means we use associated type to determine how plonk challenges are generated; and will be a bit problematic for Groth16 etc.

OTOH, between `FSRng` and `Transcript` I like the latter better. (just a personal preference, no particular reasons)

Reply:
hmmm can we use `PlonkKzgSnark`'s the associated field `Transcript` to specify the `transcript` that we are going to use.

Reply:
Oh I see, something [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6e37b93c19996497f1c1af7f9ebd527a). Got it! make sense.

Reply:
I run into another problem. Our `PlonkKzgSnark` now has to decide on one concrete Transcript struct, instead of being generic over `trait PlonkTranscript`, which means we need a SNARK impl for each type of PlonkTranscripts....

Reply:
I think I find a way out, still WIP, but it would probably be a significant simplification to our code, I'm stripping away a lot of the generic trait bound of `<E, F, P, T>` to just `<E>`.